src/sources/syslog.rs

```diff
@@ -219,59 +225,163 @@ impl SyslogDecoder {
         // |
         // | ASCII decimal number of unknown length
 
-        if let Some(i) = src.iter().position(|&b| b == b' ') {
-            let len: usize = std::str::from_utf8(&src[..i])
-                .map_err(|_| ())
-                .and_then(|num| num.parse().map_err(|_| ()))
-                .map_err(|_| {
-                    LinesCodecError::Io(io::Error::new(
-                        io::ErrorKind::InvalidData,
-                        "Unable to decode message len as number",
-                    ))
-                })?;
-
-            let from = i + 1;
-            let to = from + len;
-
-            if let Some(msg) = src.get(from..to) {
-                let s = std::str::from_utf8(msg)
-                    .map_err(|_| {
-                        LinesCodecError::Io(io::Error::new(
+        let space_pos = src.iter().position(|&b| b == b' ');
+
+        // If we are discarding, discard to the next newline.
+        let newline_pos = src.iter().position(|&b| b == b'\n');
+
+        match (
+            self.is_discarding,
+            self.discard_chars,
+            newline_pos,
+            space_pos,
+        ) {
+            (true, Some(chars), _, _) if src.len() >= chars => {
+                // We have a certain number of chars to discard.
+                // There are enough chars in this frame to discard
+                src.advance(chars);
+                self.is_discarding = false;
+                self.in_octet = false;
+                self.discard_chars = None;
+                Err(LinesCodecError::Io(io::Error::new(
+                    io::ErrorKind::Other,
+                    "Frame length limit exceeded",
+                )))
+            }
+
+            (true, Some(chars), _, _) => {
+                // We have a certain number of chars to discard.
+                // There aren't enough in this frame so we need to discard
+                // The entire frame and adjust the amount to discard accordingly.
+                self.discard_chars = Some(src.len() - chars);
+                src.advance(src.len());
+                Ok(None)
+            }
+
+            (true, _, Some(offset), _) => {
+                // When discarding we keep discarding to the next newline.
+                src.advance(offset + 1);
+                self.is_discarding = false;
+                self.in_octet = false;
+                Err(LinesCodecError::Io(io::Error::new(
+                    io::ErrorKind::Other,
+                    "Frame length limit exceeded",
+                )))
+            }
+
+            (true, _, None, _) => {
```

Comment:
I'd tend to want the second `None` made explicit here rather than relying on the ordering of the cases. That is reserve the `_` for when we don't care, and put actual values for all other places.

Of course, using a state enum eliminates at least some of these.